### PR TITLE
Common API Between Flask + Blueprint

### LINKED ISSUE #3215: Identify the common API between Flask and Blueprint classes
Blueprints are supposed to provide a very similar API to the Flask object itself, the difference being that Blueprints defer registration until later. Currently, blueprints can only be registered on apps, they're not nestable, and apps can't be registered with other apps, but this is something that #593 and #1548 want to change.

`Blueprint` and `Flask` do derive from `helpers._PackageBoundObject`, but this is mostly concerned with managing static files, templates, and other resource files. I think identifying the common API between the two and extracting it to a common base class may be a good step towards understanding how we can support app and blueprint nesting. Hopefully, it may at least make the code cleaner and easier to maintain.

### PR DESCRIPTION
**Pull request by YKo20010**
Resolves #3215

### CODE REVIEW

**Comment by YKo20010** [comment]
From the previous draft PR on our experimental branch:
image

**Comment by davidism** [comment]
`Skeleton`, `Flask`, and `Blueprint` should use `setupmethod` consistently. The fact that blueprints weren't using it (or some form of it) is probably an oversight. The point of `setupmethod` is to ensure those methods aren't called after the application starts because it would lead to inconsitent results. That seems even more true for blueprints, where adding a route, etc. to a blueprint after it's registered means it won't have been added to the app.

**Comment by davidism** [comment]
Since `setupmethod` is specific to `Flask` right now (uses `debug` and `_got_first_request`, which are `Flask`-only), this could be extracted to a helper method that both `Flask` and `Blueprint` implement.

```python
def setupmethod(f):
    @wraps(f)
    def wrapped(self, *args, **kwargs):
        if self._is_setup_finished():
            raise AssertionError(...)

        return f(*args, **kwargs)

    return wrapper

class Skeleton:
    def _is_setup_finished(self):
        raise NotImplementedError

class Flask:
    def _is_setup_finished(self):
        return self.debug and self._got_first_request

class Blueprint:
    def _is_setup_finished(self):
        return self.warn_on_modifications and self._got_registered_once
```

**Comment by davidism** [comment]
This does change the behavior a bit, since `Blueprint.warn_on_modifications` shows a warning instead of raising an error. For some reason, `warn_on_modifications` is disabled by default as well, and I'm not really clear why an `AssertionError` was used instead of a `RuntimeError`. I'll address that in a later issue though.

**Comment by davidism** [comment]
Regarding naming, a few of us in Discord liked the name `Scaffold`, it fits with Pallets and `Blueprint`. I'll let it sit a while longer, no need to rename it yet.

**Comment by chrisngyn** [comment]
I saw the messages in the Discord about switching from `skeleton.py` and `class Skeleton` to `scaffold.py` and `class Scaffold` respectively, just scrolled up and saw your comment about not needing to change it just yet -- sorry about that! 😅 We can always revert if we decide against it. 👍

File: src/flask/app.py [hunk 1]
```diff
@@ -371,7 +371,7 @@ def __init__(
         instance_relative_config=False,
         root_path=None,
     ):
-        Skeleton.__init__(
+        Scaffold.__init__(
```
Comment (davidism) [inline src/flask/app.py:374]: Use `super()`

File: src/flask/blueprints.py [hunk 2]
```diff
@@ -167,7 +167,7 @@ def __init__(
         root_path=None,
         cli_group=_sentinel,
     ):
-        Skeleton.__init__(
+        Scaffold.__init__(
```
Comment (davidism) [inline src/flask/blueprints.py:170]: Use `super()`

File: src/flask/blueprints.py [hunk 3]
```diff
@@ -258,17 +293,16 @@ def register(self, app, options, first_registration=False):
...
-    def route(self, rule, **options):
-        """Like :meth:`Flask.route` but for a blueprint.  The endpoint for the
-        :func:`url_for` function is prefixed with the name of the blueprint.
-        """
-
-        def decorator(f):
-            endpoint = options.pop("endpoint", f.__name__)
-            self.add_url_rule(rule, endpoint, f, **options)
-            return f
+    # def route(self, rule, **options):
```
Comment (davidism) [inline src/flask/blueprints.py:296]: leftover commented code

**Comment by davidism** [comment]
Moved a couple minor things into `scaffold`, added a changelog.

**Pull request merged**